pkg/tfcli/client.go

```diff
@@ -0,0 +1,316 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tfcli
+
+import (
+	"bytes"
+	"context"
+	"crypto/md5"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"text/template"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"go.uber.org/multierr"
+
+	"github.com/crossplane-contrib/terrajet/pkg/process"
+	cliErrors "github.com/crossplane-contrib/terrajet/pkg/tfcli/errors"
+	"github.com/crossplane-contrib/terrajet/pkg/tfcli/model"
+	"github.com/crossplane-contrib/terrajet/pkg/tfcli/templates"
+)
+
+const (
+	tplMain         = "main.tf.json"
+	fileTFStateLock = ".terraform.tfstate.lock.info"
+	fileState       = "terraform.tfstate"
+	fileStore       = ".store"
+	pathTerraform   = "terraform"
+	// error messages
+	errDestroyProcess   = "failed to kill Terraform CLI process"
+	errDestroyWorkspace = "failed to destroy Terraform workspace"
+	errNoProcessState   = "failed to store process state: no process state"
+	errStore            = "failed to store process state"
+	errCheckExitCode    = "failed to check process exit code"
+	errNoPlan           = "plan line not found in Terraform CLI output"
+	errPlan             = "failed to parse the Terraform plan"
+	fmtErrPath          = "failed to check path on filesystem: %s: Expected a dir: %v, found a dir: %v"
+	fmtErrLoadState     = "failed to load state from file: %s"
+	fmtErrStoreState    = "failed to store state into file: %s"
+	fmtErrAsyncRun      = "failed to run async Terraform pipeline %q with args: %v: in dir: %s"
+	fmtErrSyncRun       = "failed to run sync Terraform pipeline %q with args: %v: in dir: %s"
+
+	regexpPlanLine = `Plan:.*([\d+]).*to add,.*([\d+]) to change, ([\d+]) to destroy`
+
+	tfMsgNonExistentResource = "Cannot import non-existent remote object"
+)
+
+type client struct {
+	state       *withState
+	provider    *withProvider
+	resource    *withResource
+	execTimeout *withTimeout
+	logger      *withLogger
+	wsPath      string
+	pInfo       *process.Info
+}
+
+// returns true if no resource is to be added according to the plan
+func tfPlanCheckAdd(log string) (bool, error) {
+	r := regexp.MustCompile(regexpPlanLine)
+	m := r.FindStringSubmatch(log)
+	if len(m) != 4 || len(m[1]) == 0 {
+		return false, errors.New(errNoPlan)
+	}
+	addCount, err := strconv.Atoi(m[1])
+	if err != nil {
+		return false, errors.Wrap(err, errPlan)
+	}
+	return addCount == 0, nil
+}
+
+func (c *client) storePipelineResult(log string) error {
+	ps := c.pInfo.GetProcessState()
+	if ps == nil {
+		return errors.New(errNoProcessState)
+	}
+	return errors.Wrap(
+		ioutil.WriteFile(filepath.Join(c.wsPath, fileStore),
+			[]byte(fmt.Sprintf("%d\n%s", ps.ExitCode(), log)), 0644), errStore)
```

Comment:
I suggest we consider using https://github.com/spf13/afero/ rather than interacting with the filesystem directly. This is primarily handy to inject fake filesystems (e.g. in memory filesystems) when testing.